contributors/design-proposals/scheduling/cluster-scoped-resources.md

```diff
@@ -0,0 +1,88 @@
+
+2. IP Addresses:
+   Nodes in a cluster can be partitioned into multiple network scopes and each network scope can have a certain number of available IPs that can be assigned to pods running on those nodes. Hence, the nodes belonging to a network scope can collectively run as many pods as the IP capacity in that scope.
+
+3. Rack storage:
+   Locally attached shared storage in a rack, which is consumable by pods on nodes within a rack
+
+4. Network Bandwidth:
+   Network bandwidth shared by pods. Depending on the network topology of a cluster, pods on multiple nodes will be sharing network bandwidth with each other. In use cases where pods need guranteed network throughput, representing bandwidth as a cluster resource is essential for scheduling such pods.
+
+## Goals
+The goal is to define mechanisms to expose and consume cluster scoped resources
+
+## Design
+
+### ClusterResource type
+```
+// pkg/api/types.go:
+
+// ClusterResourceQuantity represents quantity of a ClusterResource
+type ClusterResourceQuantity struct {
+	Quantity resource.Quantity `json:"quantity"`
+	// NodeSelector is a label query over nodes which collectively provide resource Quantity
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+}
+
+// ClusterResource represents a resource which is available at a cluster level
+type ClusterResource struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+	// +optional
+	Status ClusterResourceStatus `json:"status,omitempty"`
+}
+
+type ClusterResourceStatus struct {
+	// Capacity represents the total quantity of ClusterResource
+	// +optional
+	Capacity []ClusterResourceQuantity `json:"capacity,omitempty"`
+	// Allocatable represents the quantity of ClusterResource that is available for scheduling
+	// +optional
+	Allocatable []ClusterResourceQuantity `json:"allocatable,omitempty"`
+}
+```
+`ClusterResourceStatus` captures the capacity and allocatable quantity for a `ClusteResource` in the form of `ClusterResourceQuantity`. `ClusterResourceQuantity` represents the quantity of a `ClusterResource` which is collectively consumable on nodes selected by `NodeSelector`.
+`NodeSelector` is a label query over the nodes, which collectively provide this `ClusterResource`. This field is optional, and if not specfied, it means that the `ClusterResource` is consumable across all nodes in the cluster.
+
+
+### Consuming ClusterResources
+
+ClusterResources are consumable by pods just like CPU and memory, by specifying it in the pod request. The scheduler should take care of the resource accounting for ClusterResources so that no more than the available amount is simultaneously allocated to Pods. The prefix used to identify a ClusterResource coule be 
+```
+pod.alpha.kubernetes.io/cluster-resource-
+```
+
+### Accounting in scheduler
+
+ClusterResources should be tracked as normal consumable resources and should be considered by the scheduler when determining if a pod can actually be scheduled
```

Comment:
Another important aspect of cluster resources which is not covered here is how to bind these resources to a chosen node during/after scheduling. A fairly complex logic is already added to scheduler to handle provisioning and binding PVs to nodes during scheduling. Similar processes may be needed for other resources, such TPUs, etc. I think that aspect should be covered by the proposal.

@vishh @jiayingz

Reply:
To cover that aspect,  I prefer the approach mentioned by @davidopp in his previous comment. The external agent/controller which exposes the available capacity of this resource can be made responsible for binding or making sure that those resources are ready to use when a pod is going to run on a node. Similarly when a pod dies, that agent needs to deallocate/unbind the corresponding resource and increment the available quantity so that it can be used for scheduling of new pods